doctr/io/pdf.py

```diff
@@ -3,44 +3,62 @@
 
+from typing import Any, Optional, List, Generator
 from pathlib import Path
-from typing import Any, List, Optional
 
 import numpy as np
 import pypdfium2 as pdfium
 
 from doctr.utils.common_types import AbstractFile
 
-__all__ = ["read_pdf"]
+__all__ = ["PdfRenderer"]
 
 
-def read_pdf(
-    file: AbstractFile,
-    scale: float = 2,
-    rgb_mode: bool = True,
-    password: Optional[str] = None,
-    **kwargs: Any,
-) -> List[np.ndarray]:
-    """Read a PDF file and convert it into an image in numpy format
+class PdfRenderer:
 
-    >>> from doctr.documents import read_pdf
-    >>> doc = read_pdf("path/to/your/doc.pdf")
+    def __init__(
+        self,
+        file: AbstractFile,
+        scale: float = 2,
+        page_indices: Optional[List[int]] = None,
+        password: Optional[str] = None,
+        rgb_mode: bool = True,
+        **kwargs: Any,
+    ):
+        """
+        Read a PDF file and convert it to images in numpy format.
+        This class behaves like an iterator with known length.
 
-    Args:
-        file: the path to the PDF file
-        scale: rendering scale (1 corresponds to 72dpi)
-        rgb_mode: if True, the output will be RGB, otherwise BGR
-        password: a password to unlock the document, if encrypted
-        kwargs: additional parameters to :meth:`pypdfium2.PdfDocument.render_to`
+        >>> from doctr.io.pdf import PdfRenderer
+        >>> doc = PdfRenderer("path/to/your/doc.pdf")
+        >>> n_pages = len(doc)
+        >>> first_page = next(doc)
+        >>> for further_page in doc:
+        >>>     do_something(further_page)
 
-    Returns:
-        the list of pages decoded as numpy ndarray of shape H x W x C
-    """
+        Args:
+            file: the path to the PDF file
+            scale: rendering scale (1 corresponds to 72dpi)
+            rgb_mode: if True, the output will be RGB, otherwise BGR
+            password: a password to unlock the document, if encrypted
+            kwargs: additional parameters to :meth:`pypdfium2.PdfDocument.render`
+        """
 
-    if isinstance(file, Path):
-        file = str(file)
+        pdf = pdfium.PdfDocument(file, password=password)
+        self._len = len(page_indices) if page_indices else len(pdf)
+        self._generator = pdf.render(
+            pdfium.PdfBitmap.to_numpy,
+            scale=scale, page_indices=page_indices, rev_byteorder=rgb_mode, **kwargs
+        )
 
-    # Rasterise pages to numpy ndarrays with pypdfium2
-    pdf = pdfium.PdfDocument(file, password=password)
-    renderer = pdf.render_to(pdfium.BitmapConv.numpy_ndarray, scale=scale, rev_byteorder=rgb_mode, **kwargs)
-    return [img for img, _ in renderer]
+    def __len__(self) -> int:
+        return self._len
+
+    def __next__(self) -> np.ndarray:
+        self._len -= 1
+        return next(self._generator)
+
+    def __iter__(self) -> Generator[np.ndarray, None, None]:
+        for page in self._generator:
+            self._len -= 1
+            yield page
```

Comment:
Better alternative
```python
while True:
    try:
        yield next(self)
    except StopIteration:
        break
```

This avoids duplication if further work needs to be done on each step.

Reply:
Looks better :+1: